src/read.cpp

```diff
@@ -38,15 +39,27 @@ CharacterVector read_lines_(List sourceSpec, List locale_, std::vector<std::stri
   LocaleInfo locale(locale_);
   Progress progressBar;
 
+  bool chunked = chunk_size != -1;
+
   R_len_t n = (n_max < 0) ? 10000 : n_max;
+
   CharacterVector out(n);
 
   R_len_t i = 0;
+  R_len_t pos = i + 1;
   for (Token t = tokenizer.nextToken(); t.type() != TOKEN_EOF; t = tokenizer.nextToken()) {
     if (progress && (i + 1) % 25000 == 0)
       progressBar.show(tokenizer.progress());
 
-    if (i >= n) {
+    if (chunked && i >= chunk_size) {
+      if (!as<Function>(callback["continue"])()) {
+        break;
+      }
+      out = Rf_xlengthgets(out, i);
+      as<Function>(callback["receive"])(out, pos);
+      pos = i + 1;
+      i = 0;
+    } else if (!chunked && i >= n) {
       if (n_max < 0) {
         // Estimate rows in full dataset
         n = (i / tokenizer.progress().first) * 1.2;
```

Comment:
This is a comment for line 62/75 below (`++i`). Shouldn't that be inside the conditional just above it, that assigns `out[i]`? Or better yet, `out[i++] = ...`. Otherwise, it seems like we're leaving garbage data in `out` everywhere we see a token that should be ignored...?